### ISSUE TITLE
Tests: Allow snippet based codefix tests

### ISSUE THREAD

**Issue by martin-strecker-sonarsource** [SonarSource/sonar-dotnet#5758]
The infrastructure for the code fix test,s lacks some essential features for code fix validation. Most notably, these are

* Snippet-based testing ([see](https://github.com/SonarSource/sonar-dotnet/blob/96fe491db1228e452cc30bb7dd5f89b2a9e4e41f/analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs#L220-L223))
* Code comment-based annotation support for offered code fix titles o annotate diagnostic locations with:
  * Assert code fix xyz is offered
  * Assert x number of code fixes are offered
  * Assert code fix xyz is not offered
  * Assert no code fix is offered

The absence of a fix is almost untestable at the moment and requires a lot of work. See e.g. commit 333d87791a5cec0835ad04e8cfae81263d70c75e):
* Create a copy of the test file as a new fix file
* Replace "Uncompliant" with "Fixed" in the annotations
* Create a test method
* Make sure the fixer fixes at least one diagnostics
* Repeat for each fixer.

### Support code fix annotations in tests

I would suggest extending `Verify` to also test for offered code fixer:

```cs
// Make sure "Verify" knows about the fixers. "Verify" does not execute the fixer but asks for the titles of the fixes.
builder.WithOfferedCodeFixes(typeof(Codefix1), typeof(CodeFix2)).Verify();
```

In tests:
```cs
//Title-based test for offered code fixes
Sut1(); // Noncompliant {{Diagnostic message}} [[CodeFix1 title 1||CodeFix1 title 2||Codefix2 title 1]]
// Number based test for offered code fixes. Exact two fixes need to be offered (no matter from which code fix).
Sut2(); // Noncompliant {{Diagnostic message}} [[=2]]
// Make sure no fix is offered
Sut2(); // Noncompliant {{Diagnostic message}} [[=0]]
```

### Allow snippet-based fixing

This has the benefit of better test isolation. Today, every change in the test file needs to be duplicated in the fixed file. Snipped-based testing would simplify the testing:

```cs
builder
  .WithCodeFix<CodeFix>() // single code fix as is
  .AddSnippet(@"someTestCode")  
  .AddFixedSnippet(@"fixedCode") // Might be restricted to single snippets in Verifier.ValidateCodeFix()
  .VerifyCodeFix();
```

Each test class could contain a small helper function like so:

```cs
TestCodeFix(string before, string after) => builder ...
```

**Comment by pavel-mikula-sonarsource** [comment]
The `WithOfferedCodeFixes` part is not needed, all code fixes should be provided by a single code fix provider.

Asserting code fix titles is a good idea.

The snippet part should also not be needed. The rest of TestCase file can be deleted to debug isolated scenario. Same as debugging isolated scenario for the rule itself. We don't want to have 1000 tests for each rule.

**Comment by Corniel** [comment]
I can relate to this. See: #6966. In this case the fix is even trickier to test as the "Replace noncompliant with fixed part" messes with the fix it self.

For #6043 I wrote a separate test case for the fix. I think that might be a good practice in most/all cases; the scenario's to cover for a fix do not relate one to one to the scenario's to cover for the rule itself. In the referenced story, the test case for the fix was way simpler.

**Comment by pavel-mikula-sonarsource** [comment]
Moved to [NET-1854](https://sonarsource.atlassian.net/browse/NET-1854)

[NET-1854]: https://sonarsource.atlassian.net/browse/NET-1854?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
